src/StreamrClient.js

```diff
@@ -268,6 +268,18 @@ export default class StreamrClient extends EventEmitter {
         // If connected, emit a publish request
         if (this.isConnected()) {
             const streamMessage = await this.msgCreationUtil.createStreamMessage(streamObjectOrId, data, timestamp, partitionKey)
+            // if happened to disconnect before completed
```

Comment:
Ooh this is getting messy! Would be great to not to have to repeat any part of the "if not connected, then queue the message and autoconnect" logic. There must be a cleaner way. How about along the lines of:

- Create the `streamMessage` **before** the first `this.isConnected()` check on line 269
- Then if we end up in the autoconnect procedure, push a function that calls `this._requestPublish(streamMessage, sessionToken)` into the publish queue instead of the function that calls `this.publish(streamId, data, timestamp, partitionKey)`

Reply:
Done in e2e9a19

@hpihkala Note one side effect of this change is that the timestamp will now match the point where `publish()` was called, rather than when the connection happened. This is probably desirable anyway. cc @harbu

Reply:
Sound reasonable imo.

Reply:
A lot cleaner now, thanks!